include/pstl_offload/internal/usm_memory_replacement_common.h

```diff
@@ -0,0 +1,176 @@
+    void* __ptr = sycl::aligned_alloc_shared(__usm_alignment, __usm_size, *__device, __context);
+
+    if (__ptr != nullptr)
+    {
+        void* __original_pointer = __ptr;
+        __ptr = static_cast<char*>(__ptr) + __base_offset;
+        __block_header* __header = static_cast<__block_header*>(__ptr) - 1;
+        *__header = __block_header{__uniq_type_const, __original_pointer, __device, __size};
+    }
+
+    return __ptr;
+}
+
+inline bool
+__same_memory_page(void* __ptr1, void* __ptr2)
+{
+    std::uintptr_t __page_size = __get_memory_page_size();
+    std::uintptr_t __page_mask = ~(__page_size - 1);
+    std::uintptr_t __ptr1_page_begin = std::uintptr_t(__ptr1) & __page_mask;
+    std::uintptr_t __ptr2_page_begin = std::uintptr_t(__ptr2) & __page_mask;
+    return __ptr1_page_begin == __ptr2_page_begin;
+}
+
+inline auto
+__get_original_realloc()
+{
+    using __realloc_func_type = void* (*)(void*, std::size_t);
+
+    static __realloc_func_type __orig_realloc = __realloc_func_type(dlsym(RTLD_NEXT, "realloc"));
+    return __orig_realloc;
+}
+
+inline void
+__free_usm_pointer(__block_header* __header)
+{
+    assert(__header != nullptr);
+    sycl::context __context = __header->_M_device->get_platform().ext_oneapi_get_default_context();
+    __header->_M_uniq_const = 0;
+    sycl::free(__header->_M_original_pointer, __context);
+}
+
+inline void*
+__realloc_real_pointer(void* __user_ptr, std::size_t __new_size)
+{
+    assert(__user_ptr != nullptr);
+    __block_header* __header = static_cast<__block_header*>(__user_ptr) - 1;
+
+    void* __result = nullptr;
+
+    if (__same_memory_page(__user_ptr, __header) && __header->_M_uniq_const == __uniq_type_const)
+    {
+        if (__header->_M_requested_number_of_bytes >= __new_size)
+        {
+            // No need to reallocate memory, previously allocated number of bytes is enough to store __new_size
+            __result = __user_ptr;
+        }
+        else
+        {
+            // Reallocate __new_size
+            assert(__header->_M_device != nullptr);
+            void* __new_ptr = __allocate_shared_for_device(__header->_M_device, __new_size, alignof(std::max_align_t));
+
+            if (__new_ptr != nullptr)
+            {
+                std::memcpy(__new_ptr, __user_ptr, __header->_M_requested_number_of_bytes);
+
+                // Free previously allocated memory
+                __free_usm_pointer(__header);
+                __result = __new_ptr;
+            }
+            else
+            {
+                errno = ENOMEM;
+            }
+        }
+    }
+    else
+    {
+        // __user_ptr is not a USM pointer, use original realloc function
+        __result = __get_original_realloc()(__user_ptr, __new_size);
+    }
+    return __result;
+}
+
+static void*
+__internal_realloc(void* __user_ptr, std::size_t __new_size)
+{
+    return __user_ptr == nullptr ? std::malloc(__new_size) : __realloc_real_pointer(__user_ptr, __new_size);
```

Comment:
Why do we use std::malloc here, and not directly our allocation function?

Reply:
According to the `realloc` documentation - if the pointer is `nullptr`, it should behave the same as `malloc`.
The only internal function that we can call here is `errno_internal_aligned_alloc(std::size_t size, std::size_t alignment)` that is used in our `malloc` implementation. It explicitly requires the alignment value that possible creates error-prone to pass the alignment different than used in `malloc`. IMHO, calling `malloc` here is more straightforward here and `std::malloc` call will find our replacement version of `malloc`.